File: ruby/triangle/triangle.rb
# frozen_string_literal: true

class Triangle
  def initialize(given_input)
    @sides = given_input
  end

  def equilateral?
    verify_triangularity_total && unique_entries == 1
  end

  def isosceles?
    verify_triangularity_total && (equilateral? || unique_entries == 2)
  end

  def scalene?
    verify_triangularity_total && unique_entries == 3
  end

  def degenerate?
    verify_triangularity_total && verify_triangular_degeneracy
  end

  private

  def unique_entries
    @sides.uniq.count
  end

  def verify_triangularity_total
    verify_triangularity_no_zero_sides && verify_triangularity_side_length
  end

  def verify_triangularity_no_zero_sides
    @sides.select(&:zero?).empty?
  end

  def verify_triangularity_side_length
    triangularity = String.new
    @sides.each do |x|
      triangularity += "#{x.to_i <= (@sides.sum.to_i - x.to_i)} "
    end
    triangularity.split.all? { |sides| sides == 'true' }
  end

  def verify_triangular_degeneracy
    triangularity = String.new
    @sides.each do |x|
      triangularity += "#{x == @sides.sum - x} "
    end
    triangularity.split.any? { |sides| sides == 'true' }
  end
end

File: prolog/triangle/triangle.pl
triangle(Arg, Arg, Arg, "equilateral") :-
  Arg > 0,
  !.

triangle(Arg1, Arg2, Arg3, "scalene") :-
  Arg1 > 0,
  Arg2 > 0,
  Arg3 > 0,
  Arg1 \== Arg2,
  Arg2 \== Arg3,
  Arg1 \== Arg3,
  Arg1 + Arg2 > Arg3,
  Arg1 + Arg3 > Arg2,
  Arg2 + Arg3 > Arg1,
  !.

triangle(Arg, Arg, Arg2, "isosceles") :-
  Arg > 0,
  Arg2 > 0,
  2 * Arg >= Arg2,
  !.

triangle(Arg, Arg2, Arg, "isosceles") :-
  Arg > 0,
  Arg2 > 0,
  2 * Arg >= Arg2,
  !.

triangle(Arg2, Arg, Arg, "isosceles") :-
  Arg > 0,
  Arg2 > 0,
  2 * Arg >= Arg2,
  !.